service/history/historyEngine.go

```diff
@@ -294,6 +302,56 @@ func (e *historyEngineImpl) GetWorkflowExecutionNextEventID(
 		RunId:      request.Execution.RunId,
 	}
 
+	response, err := e.getWorkflowExecutionNextEventID(domainID, execution)
+	if err != nil {
+		return nil, err
+	}
+
+	// expectedNextEventID is 0 when caller want to get the current next event ID without blocking
+	expectedNextEventID := request.GetExpectedNextEventId()
+
+	// if caller decide to long poll on workflow execution
+	// and the event ID we are looking for is smaller than current next event ID
+	if expectedNextEventID >= response.GetEventId() && response.GetIsWorkflowRunning() {
+		subscriberID, channel, err := e.historyEventNotifier.WatchHistoryEvent(&domainID, &execution)
+		if err != nil {
+			return nil, err
+		}
+		defer e.historyEventNotifier.UnwatchHistoryEvent(&domainID, &execution, subscriberID)
+
+		// check again in case the next event ID is updated
+		response, err = e.getWorkflowExecutionNextEventID(domainID, execution)
+		if err != nil {
+			return nil, err
+		}
+
+		if expectedNextEventID < response.GetEventId() || !response.GetIsWorkflowRunning() {
+			return response, nil
+		}
+
+		timeoutChan := time.NewTimer(e.shard.GetConfig().LongPollExpirationInterval).C
+		for {
+			select {
+			case event := <-channel:
+				response.EventId = common.Int64Ptr(event.nextEventID)
+				response.IsWorkflowRunning = common.BoolPtr(event.isWorkflowRunning)
+				if expectedNextEventID < response.GetEventId() || !response.GetIsWorkflowRunning() {
+					return response, nil
+				}
+			case <-timeoutChan:
+				return response, nil
+			case <-ctx.Done():
+				return nil, &workflow.BadRequestError{Message: "timeout: frontend timeout the request"}
```

Comment:
BadRequest does not seems like the right error.  You should call ctx.GetErr() to get the actual error.

Reply:
ok make sense.

i was referencing:
https://github.com/uber/cadence/blob/9db7677252ec531e2f493aa28d7cc1a2c89463bd/service/matching/matchingEngine.go#L404